File: LeetCode/0149/golang/lc0149.go
package lc0149

type Point struct {
	x int
	y int
}

// Runtime complexity: O(log(max(x, y)))
// Auxiliary space: O(1)
func gcd(x int, y int) int {
	for y != 0 {
		x, y = y, x%y
	}
	return x
}

// Runtime complexity: O(1)+O(gcd()) == O(log(max(x, y)))
// Auxiliary space: O(1)
func getSlope(a Point, b Point) Point {
	dx, dy := a.x-b.x, a.y-b.y
	if dy == 0 {
		return Point{a.y, 0}
	}
	if dx == 0 {
		return Point{0, a.x}
	}
	d := gcd(dx, dy)
	return Point{dx / d, dy / d}
}

// Runtime complexity: O(n^2)*O(getSlope), which is still O(n^2) in the end.
// Auxiliary space: O(n)
// Subjective level: hard
// Solved on: 2026-01-30
func maxPoints(points [][]int) int {
	ans := 0
	for i, point := range points {
		slopeToCount := map[Point]int{}
		numRepeatedPoints := 1
		numSameSlopePoints := 0
		p1 := Point{point[0], point[1]}
		for j := i + 1; j < len(points); j++ {
			point2 := points[j]
			p2 := Point{point2[0], point2[1]}
			if point[0] == point2[0] && point[1] == point2[1] {
				numRepeatedPoints += 1
			} else {
				slope := getSlope(p1, p2)
				_, isPresent := slopeToCount[slope]
				if isPresent {
					slopeToCount[slope] += 1
				} else {
					slopeToCount[slope] = 1
				}

				numSameSlopePoints = max(numSameSlopePoints, slopeToCount[slope])
			}
		}

		ans = max(ans, numRepeatedPoints+numSameSlopePoints)
	}
	return ans
}

File: LeetCode/0149/python/lc0149.py
import collections
from typing import List

# No need; using tuple[int, int] for now.
# class Point:
#    x: int
#    y: int

def recursive_gcd(x: int, y: int) -> int:
    if y == 0:
        return x
    return recursive_gcd(y, x%y)

# Runtime complexity: O(log(max(x, y)))
# Auxiliary space: O(1)
def gcd(x: int, y: int) -> int:
    while y != 0:
        x, y = y, x%y
    return x

# Runtime complexity: O(1)+O(gcd()) == O(log(max(x, y)))
# Auxiliary space: O(1)
# def getSlope(p1: List[int], p2: List[int]) -> List[int]:
def getSlope(p1: List[int], p2: List[int]) -> tuple[int, int]:
    dx, dy = p1[0]-p2[0], p1[1]-p2[1]
    if dx == 0:
        return (0, p1[0])
    if dy == 0:
        return (p1[1], 0)
    g = gcd(dx, dy)
    return (dx//g, dy//g)

class Solution:
    # Runtime complexity: O(n^2)*O(getSlope), which is still O(n^2) in the end.
    # Auxiliary space: O(n)
    # Subjective level: hard
    # Solved on: 2026-01-30
    def maxPoints(self, points: List[List[int]]) -> int:
        ans = 0
        for i1, p1 in enumerate(points):
            slopeToCount = collections.defaultdict(int)
            numRepeatedPoints = 1
            numSameSlopePoints = 0
            for i2 in range(i1+1, len(points)):
                p2 = points[i2]
                if p1 == p2:
                    numRepeatedPoints += 1
                else:
                    slope = getSlope(p1, p2)
                    slopeToCount[slope] += 1
                    numSameSlopePoints = max(numSameSlopePoints, slopeToCount[slope])
            ans = max(ans, numRepeatedPoints + numSameSlopePoints)
        return ans
